### PR TITLE
Core/Movement:

### LINKED ISSUE #5086: Warsong Gulch Interior/Exterior Boundary is Incorrect (Doesn't Allow Mounting) [$10]
Warsong Gulch Interior/Exterior Boundary is Incorrect (Doesn't Allow Mounting)

## CURRENT BEHAVIOUR:

You must be 5+ yards outside of the building to mount up, causing people to awkwardly stand and click and try to find a spot outside they are allowed to mount at.

As you can see I'm quite far outside but I can't mount due to being indoors. "Can only use outside" is the error presented.

WoWScrnShot_033121_211327

## EXPECTED BLIZZLIKE BEHAVIOUR:

Should be able to mount when outdoors.

## SOURCE:

https://youtu.be/UyFATw5m7uE?t=86

Observe how the entire team gets an inch out of the door and begins mounting at 1 min 25 secs ish. Video is timestamped.

## STEPS TO REPRODUCE THE PROBLEM:

1. Enter WSG
2. Run out as the game begins and try to mount
3. Run in/ou/in/out and observe mount behaviour.

## AC HASH/COMMIT:

AzerothCore rev. 8bacb39d5334 2021-03-20 

## MODULES:

CFBG

## CUSTOMIZATIONS:

I do use some texture patches but texture patches shouldn't affect the size/shape of the map right?

<bountysource-plugin>

---
There is a **[$10 open bounty](https://www.bountysource.com/issues/97541657-warsong-gulch-interior-exterior-boundary-is-incorrect-doesn-t-allow-mounting?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by UltraNix**
Get zone/area IDs from vmap data in the liquid update.
Add new method Map::getFullVMapDataForPosition to get area info and liquid info in a single vmap lookup
Adjust GetZoneId/GetAreaId on WorldObject to always return these cached fields.
Clean up liquid state handling on Unit and Player
Implemented getting area id from gameobject spawns.
Removed old core related to getting movement flags dependent on environment.
Movement flags are now processed more precisely and dynamically.
Original source: TrinityCore.
Fixed #5086
Updated #2208.

## Changes Proposed:
-  Backported changes from TrinityCore about getting cached data (areaId, zoneId, bolean outdoor and liquid data) when unit/object is moving
-  Cleanup liquid data code.
- Removed old and not smooth `UpdateEnvironmentIfNeeded` and everything related to it.

## Issues Addressed:

- Closes #5086.

## Tests Performed:

- Tested ingame in open world with players, flying mobs, mounts, swiming, etc

## How to Test the Changes:

There is a list what to test:
1. Spells/items that should be usable only/forbidden outdoor/indoor (e.g. mounts)
2. Test breathing timer/bar while swimming/entering/leaving water/lava/ocean.
3. Test proper area/zone ids by typiing ingame `.gps` in different locations all around the world.
4. Test flying/flyable mobs/NPCs and their transition between flying/walking on the ground.

The best spots to test it is IcecrownCitadel and cities.

Elune changes [here](https://github.com/ElunaLuaEngine/Eluna/pull/377)

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by UltraNix** [comment]
Fixes all the flying mobs that still have flying animation when landing/walking on the ground.

**Comment by FrancescoBorzi** [comment]
@azerothcore/testers @azerothcore/senior-bug-triaging @azerothcore/developers please check this PR

File: src/server/game/Entities/Object/Object.h [hunk 1]
```diff
@@ -811,9 +806,11 @@ class WorldObject : public Object, public WorldLocation
...
     [[nodiscard]] bool InSamePhase(uint32 phasemask) const { return m_useCombinedPhases ? GetPhaseMask() & phasemask : GetPhaseMask() == phasemask; }
 
-    [[nodiscard]] virtual uint32 GetZoneId(bool forceRecalc = false) const;
-    [[nodiscard]] virtual uint32 GetAreaId(bool forceRecalc = false) const;
-    virtual void GetZoneAndAreaId(uint32& zoneid, uint32& areaid, bool forceRecalc = false) const;
+    [[nodiscard]] uint32 GetZoneId() const { return _zoneId; }
+    [[nodiscard]] uint32 GetAreaId() const { return _areaId; }
+    void GetZoneAndAreaId(uint32& zoneid, uint32& areaid) const { zoneid = _zoneId, areaid = _areaId; }
+    [[nodiscard]] bool IsOutdoors() const { return _outdoors; }
+    LiquidData const& GetLiquidData() const { return _liquidData; }
```
Comment (pklloveyou) [inline src/server/game/Entities/Object/Object.h:813]:
```
    [[nodiscard]] uint32 GetZoneId() const;
    [[nodiscard]] uint32 GetAreaId() const;
    void GetZoneAndAreaId(uint32& zoneid, uint32& areaid) const;
    [[nodiscard]] bool IsOutdoors() const;
    LiquidData const& GetLiquidData() const;
```

**Comment by pklloveyou** [comment]
In many outdoor places, riding automatically unloads the mount

**Comment by UltraNix** [comment]
@pklloveyou Give me the exact coords.
It looks okay on my local.

**Comment by pklloveyou** [comment]
It may have something to do with my anti cheating module O.o

**Comment by FrancescoBorzi** [comment]
@UltraNix the build is failing and there are conflicts

**Comment by Azcobu** [comment]
Checking this vs an old ticket about mounts being useable indoors in Grom'gol Base Camp's inn (https://github.com/azerothcore/azerothcore-wotlk/issues/6640) and found this issue is still occurring:

WoWScrnShot_071821_220444

But I don't know enough about this problem to say if these changes are supposed to fix this, as it could be a problem with this particular inn. But it is an instance of a forbidden indoor item working when it shouldn't.

**Comment by UltraNix** [comment]
#6093 is not handled by this PR.
#6640 is not a bug.

**Comment by Azcobu** [comment]
Would this PR affect NPC transitions between land and water as well? We had some issues like that (https://github.com/azerothcore/azerothcore-wotlk/issues/6541, https://github.com/azerothcore/azerothcore-wotlk/issues/6171) and someone said it might be the land/water transition. Would you expect this PR to solve those, or wouldn't it apply?

**Comment by UltraNix** [comment]
@Azcobu Might be another problem (with missing swimming flag or spline run), but you can test it.

**Comment by mpfans** [comment]
I will test it in the near future

**Comment by FrancescoBorzi** [comment]
> I will test it in the near future

thanks @mpfans , please let us know

**Comment by Mitradis** [comment]
it helps with https://github.com/azerothcore/azerothcore-wotlk/issues/7359 ?

**Comment by FrancescoBorzi** [comment]
@UltraNix  there are conflicts

**Comment by FrancescoBorzi** [comment]
@UltraNix build failing

**Comment by Helias** [comment]
```
In file included from /home/runner/work/azerothcore-wotlk/azerothcore-wotlk/modules/mod-eluna-lua-engine/LuaEngine/LuaFunctions.cpp:35:
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/modules/mod-eluna-lua-engine/LuaEngine/MapMethods.h:198:46: fatal error: too few arguments to function call, expected 4, have 3
        Eluna::Push(L, map->GetAreaId(x, y, z));
                       ~~~~~~~~~~~~~~        ^
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/server/game/Maps/Map.h:378:19: note: 'GetAreaId' declared here
    [[nodiscard]] uint32 GetAreaId(uint32 phaseMask, float x, float y, float z) const;
                  ^
1 error generated.
```

still failing :(

**Comment by UltraNix** [comment]
@Helias Look here: https://github.com/ElunaLuaEngine/Eluna/pull/377

File: src/common/Collision/DynamicTree.h [hunk 2]
```diff
@@ -32,6 +37,9 @@ class DynamicMapTree
     bool GetIntersectionTime(uint32 phasemask, const G3D::Ray& ray,
                              const G3D::Vector3& endPos, float& maxDist) const;
 
+    bool getAreaInfo(float x, float y, float& z, uint32 phasemask, uint32& flags, int32& adtId, int32& rootId, int32& groupId) const;
+    void getAreaAndLiquidData(float x, float y, float z, uint32 phasemask, uint8 reqLiquidType, VMAP::AreaAndLiquidData& data) const;
```
Comment (Kitzunu) [inline src/common/Collision/DynamicTree.h:41]:
```suggestion
    bool GetAreaInfo(float x, float y, float& z, uint32 phasemask, uint32& flags, int32& adtId, int32& rootId, int32& groupId) const;
    void GetAreaAndLiquidData(float x, float y, float z, uint32 phasemask, uint8 reqLiquidType, VMAP::AreaAndLiquidData& data) const;
```

File: src/common/Collision/Management/IVMapManager.h [hunk 3]
```diff
@@ -79,8 +105,10 @@ namespace VMAP
         Query world model area info.
         \param z gets adjusted to the ground height for which this are info is valid
         */
-        virtual bool GetAreaInfo(unsigned int pMapId, float x, float y, float& z, uint32& flags, int32& adtId, int32& rootId, int32& groupId) const = 0;
-        virtual bool GetLiquidLevel(uint32 pMapId, float x, float y, float z, uint8 ReqLiquidType, float& level, float& floor, uint32& type) const = 0;
+        virtual bool GetAreaInfo(uint32 pMapId, float x, float y, float& z, uint32& flags, int32& adtId, int32& rootId, int32& groupId) const      = 0;
+        virtual bool GetLiquidLevel(uint32 pMapId, float x, float y, float z, uint8 ReqLiquidType, float& level, float& floor, uint32& type, uint32& mogpFlags) const = 0;
+        // get both area + liquid data in a single vmap lookup
+        virtual void getAreaAndLiquidData(uint32 mapId, float x, float y, float z, uint8 reqLiquidType, AreaAndLiquidData& data) const = 0;
```
Comment (Kitzunu) [inline src/common/Collision/Management/IVMapManager.h:111]: same as above

File: src/common/Collision/Models/GameObjectModel.cpp [hunk 4]
```diff
@@ -187,6 +189,69 @@ bool GameObjectModel::intersectRay(const G3D::Ray& ray, float& MaxDist, bool Sto
     return hit;
 }
 
+void GameObjectModel::intersectPoint(G3D::Vector3 const& point, VMAP::AreaInfo& info, uint32 ph_mask) const
```
Comment (Kitzunu) [inline src/common/Collision/Models/GameObjectModel.cpp:192]: same as above

File: src/common/Collision/Models/GameObjectModel.h [hunk 5]
```diff
@@ -53,8 +55,12 @@ class GameObjectModel
     void enable(uint32 ph_mask) { phasemask = ph_mask; }
 
     [[nodiscard]] bool isEnabled() const { return phasemask != 0; }
+    [[nodiscard]] bool isMapObject() const { return isWmo; }
```
Comment (Kitzunu) [inline src/common/Collision/Models/GameObjectModel.h:58]: same as above

File: src/server/game/Entities/Creature/Creature.cpp [hunk 6]
```diff
@@ -3016,6 +3018,84 @@ float Creature::GetAggroRange(Unit const* target) const
...
+        return;
+
+    // Creatures with CREATURE_FLAG_EXTRA_NO_MOVE_FLAGS_UPDATE should control MovementFlags in your own scripts
+    if (info->flags_extra & CREATURE_FLAG_EXTRA_NO_MOVE_FLAGS_UPDATE)
+        return;
+
+    float z = GetPositionZ();
+    float ground = GetFloorZ();
+
+    bool IsInAir = false;
```
Comment (Kitzunu) [inline src/server/game/Entities/Creature/Creature.cpp:3038]:
```suggestion
    bool isInAir = false;
```

**Review by Kitzunu — CHANGES_REQUESTED** [review]

**Review by Mitradis — APPROVED** [review]

**Pull request merged**